pgx/src/spi.rs

```diff
@@ -378,8 +378,125 @@ impl SpiClient {
+            argtypes.push(argtype.value());
+
+            match datum {
+                Some(datum) => {
+                    // ' ' here means that the datum is not null
+                    datums.push(datum);
+                    nulls.push(' ' as std::os::raw::c_char);
+                }
+
+                None => {
+                    // 'n' here means that the datum is null
+                    datums.push(0);
+                    nulls.push('n' as std::os::raw::c_char);
+                }
+            }
+        }
+
+        let ptr = unsafe {
+            pg_sys::SPI_cursor_open_with_args(
+                std::ptr::null_mut(), // let postgres assign a name
+                src.as_ptr(),
+                nargs as i32,
+                argtypes.as_mut_ptr(),
+                datums.as_mut_ptr(),
+                nulls.as_ptr(),
+                false,
+                0,
+            )
+        };
+        SpiCursor { ptr }
+    }
+
+    /// find a cursor in transaction by name. Name should usually be retrieved via [`SpiCursor::into_name`]
+    pub fn find_cursor(
+        &mut self,
+        name: String,
+    ) -> SpiCursor {
+        use pgx_pg_sys::AsPgCStr;
+
+        let ptr = unsafe {
+            pg_sys::SPI_cursor_find(name.as_pg_cstr())
+        };
+        if ptr.is_null() {
+            error!("cursor named \"{}\" not found", name);
+        }
+        SpiCursor { ptr }
+    }
+}
+
+
+pub struct SpiCursor {
+    ptr: pg_sys::Portal,
+}
+
+impl SpiCursor {
+    /// fetch some rows from a cursor
+    pub fn fetch(
+        &mut self,
+        count: i64
+    ) -> SpiTupleTable {
+        unsafe {
+            pg_sys::SPI_tuptable = std::ptr::null_mut();
+        }
+        unsafe {
+            pg_sys::SPI_cursor_fetch(
+                self.ptr,
+                true,
+                count,
+            )
+        }
+        SpiTupleTable {
+            status_code: SpiOk::Fetch,
+            table: unsafe { pg_sys::SPI_tuptable },
+            size: unsafe { pg_sys::SPI_processed as usize },
+            tupdesc: if unsafe { pg_sys::SPI_tuptable }.is_null() {
+                None
+            } else {
+                Some(unsafe { (*pg_sys::SPI_tuptable).tupdesc })
+            },
+            current: -1,
+        }
+    }
+
+    /// consumes the cursor and returns its name.
+    /// This allows to fetch it in a later SPI session within the same transaction
+    /// using [`SpiClient::find_cursor()`]
+    pub fn into_name(self) -> String {
+        unsafe { std::ffi::CStr::from_ptr((*self.ptr).name) }
+            .to_str().expect("non-utf8 cursor name").to_string()
+    }
+
+    /// closes the cursor and releases its resources.
+    /// Not mandatory, as all open cursors are closed automatically at the end of a transaction.
+    pub fn close(mut self) {
+        let ptr = std::mem::replace(&mut self.ptr, std::ptr::null_mut());
+        unsafe {
+            pg_sys::SPI_cursor_close(ptr)
```

Comment:
I wonder if this should be part of `Drop`...

Reply:
Leaving cursors open to be retrieved by name later in transaction is a core use case, so always closing on `Drop` would not work.

I experimented with having `SpiCursor::into_name()` invalidating the inner `ptr`, and closing conditionally in `Drop` in case `into_inner` hadn't been called, so that if you did not retrieve the name, it would be closed, since you would not have a handle for `find_cursor` later. 
Unfortunately then you would need to call `into_name` _every_time_ you retrieved it via `find_cursor`, or else it would be closed on Drop.

Of course this could be handled by having a `close_on_drop` flag, and setting it to false for cursors retrieved by name, but it looked like too much complexity for little gain.

For now I've tried to make the doc clearer on `close`, but I'm open to better ideas.